src/Algebra/Graph/Relation/Symmetric.hs

```diff
@@ -13,25 +13,509 @@ module Algebra.Graph.Relation.Symmetric (
+-- @
+-- forest []                                                  == 'empty'
+-- forest [x]                                                 == 'tree' x
+-- forest [Node 1 [Node 2 [], Node 3 []], Node 4 [Node 5 []]] == 'edges' [(1,2), (1,3), (2,1), (3,1), (4,5), (5,4)]
+-- forest                                                     == 'overlays' . 'map' 'tree'
+-- @
+forest :: Ord a => Tree.Forest a -> SymmetricRelation a
+forest = overlays . map tree
+
+-- | Remove a vertex from a given graph.
+-- Complexity: /O(n + m)/ time.
+--
+-- @
+-- removeVertex x ('vertex' x)       == 'empty'
+-- removeVertex 1 ('vertex' 2)       == 'vertex' 2
+-- removeVertex x ('edge' x x)       == 'empty'
+-- removeVertex 1 ('edge' 1 2)       == 'vertex' 2
+-- removeVertex x . removeVertex x   == removeVertex x
+-- @
+removeVertex :: Ord a => a -> SymmetricRelation a -> SymmetricRelation a
+removeVertex x = SymmetricRelation . R.removeVertex x . fromSymmetric
+
+-- | Remove an edge from a given graph.
+-- Complexity: /O(log(m))/ time.
+--
+-- @
+-- removeEdge x y ('AdjacencyMap.edge' x y)       == 'vertices' [x,y]
+-- removeEdge x y . removeEdge x y == removeEdge x y
+-- removeEdge x y == removeEdge y x
+-- removeEdge x y . 'removeVertex' x == 'removeVertex' x
+-- removeEdge 1 1 (1 * 1 * 2 * 2)  == 1 * 2 * 2
+-- removeEdge 1 2 (1 * 1 * 2 * 2)  == 1 * 1 + 2 * 2
+-- @
+removeEdge :: Ord a => a -> a -> SymmetricRelation a -> SymmetricRelation a
+removeEdge x y r = let (RI.Relation d rr) = fromSymmetric r
+                       in SymmetricRelation $ RI.Relation d (Set.delete (y, x) . Set.delete (x, y) $ rr)
+
+-- | The function @'replaceVertex' x y@ replaces vertex @x@ with vertex @y@ in a
+-- given 'AdjacencyMap'. If @y@ already exists, @x@ and @y@ will be merged.
+-- Complexity: /O((n + m) * log(n))/ time.
+--
+-- @
+-- replaceVertex x x            == id
+-- replaceVertex x y ('vertex' x) == 'vertex' y
+-- replaceVertex x y            == 'mergeVertices' (== x) y
+-- @
+replaceVertex :: Ord a => a -> a -> SymmetricRelation a -> SymmetricRelation a
+replaceVertex u v = gmap $ \w -> if w == u then v else w
+
+-- | Merge vertices satisfying a given predicate into a given vertex.
+-- Complexity: /O((n + m) * log(n))/ time, assuming that the predicate takes
+-- /O(1)/ to be evaluated.
+--
+-- @
+-- mergeVertices ('const' False) x    == id
+-- mergeVertices (== x) y           == 'replaceVertex' x y
+-- mergeVertices 'even' 1 (0 * 2)     == 1 * 1
+-- mergeVertices 'odd'  1 (3 + 4 * 5) == 4 * 1
+-- @
+mergeVertices :: Ord a => (a -> Bool) -> a -> SymmetricRelation a -> SymmetricRelation a
+mergeVertices p v = gmap $ \u -> if p u then v else u
+
+-- | Transpose a given graph.
+-- Complexity: /O(1)/ time.
+--
+-- @
+-- transpose 'empty'       == 'empty'
+-- transpose ('vertex' x)  == 'vertex' x
+-- transpose ('edge' x y)  == 'edge' y x
+-- transpose  == id
+-- 'edgeList' . transpose  == 'edgeList'
+-- @
+transpose :: SymmetricRelation a -> SymmetricRelation a
+transpose = id
```

Comment:
Maybe we shouldn't provide this function at all.

Reply:
Okay!